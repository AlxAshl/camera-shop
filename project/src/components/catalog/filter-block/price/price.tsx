import { useEffect, useState } from 'react';
import { useSearchParams } from 'react-router-dom';
import { useAppSelector } from '../../../../hooks/useAppSelector';
import { ProductType } from '../../../../types/product';
import useDebounce from '../../../../hooks/use-debounce';
import { getProductsByPrice } from '../../../../store/product-process/selectors';
import { URLParams } from '../../../../const';


export function Price() {

  const [inputValue, setInputValue] = useState({ min: '', max: ''});
  const [searchParams, setSearchParams] = useSearchParams();
  const productsByPrice: ProductType[] = useAppSelector(getProductsByPrice);
  const debouncedMinInput = useDebounce(inputValue.min, 1250);
  const debouncedMaxInput = useDebounce(inputValue.max, 1250);

  useEffect(
    () => {

      if (debouncedMinInput !== '' && productsByPrice) {
        if (Number(debouncedMinInput) < productsByPrice[0].price && inputValue.min !== '') {
          setInputValue((prev) => ({...prev, min: String(productsByPrice[0].price)}));
          searchParams.delete(URLParams.PriceMin);
          searchParams.append(URLParams.PriceMin, String(productsByPrice[0].price));
        }
        else {
          if (Number(debouncedMinInput) > Number(inputValue.max) && inputValue.max !== '') {
            setInputValue((prev) => ({...prev, min: inputValue.max}));
            searchParams.delete(URLParams.PriceMin);
            searchParams.append(URLParams.PriceMin, inputValue.max);
          }
          if(Number(inputValue.min) > productsByPrice[productsByPrice.length - 1].price) {
            setInputValue((prev) => ({...prev, min: String(productsByPrice[productsByPrice.length - 1].price)}));
            searchParams.delete(URLParams.PriceMin);
            searchParams.append(URLParams.PriceMin, String(productsByPrice[productsByPrice.length - 1].price));
          }
          if(!productsByPrice.find((product) => product.price === Number(inputValue.min))) {
            const newPriceInput = productsByPrice.find((product, i) => {
              const prevProduct = productsByPrice[i - 1];
              return prevProduct && (prevProduct.price < Number(inputValue.min)) && (product.price > Number(inputValue.min));
            });
            const newMinPriceInput = newPriceInput?.price;
            setInputValue((prev) => ({...prev, min: String(newMinPriceInput)}));
            searchParams.delete(URLParams.PriceMin);
            searchParams.append(URLParams.PriceMin, String(newMinPriceInput));
          }
        }
      }

      else {
        const priceMinParam = searchParams.get(URLParams.PriceMin) || '';
        if (priceMinParam !== inputValue.min) {
          setInputValue((prev) => ({...prev, min: priceMinParam}));
        }
      }
      setSearchParams(searchParams);
    },
    [debouncedMinInput, searchParams]
  );

  useEffect(
    () => {

      if (debouncedMaxInput !== '' && productsByPrice) {
        if (Number(inputValue.max) > productsByPrice[productsByPrice.length - 1].price && inputValue.max !== '') {
          setInputValue((prev) => ({...prev, max: String(productsByPrice[productsByPrice.length - 1].price)}));
        }
        else {
          if (Number(debouncedMaxInput) < Number(inputValue.min)) {
            setInputValue((prev) => ({...prev, max: inputValue.min}));
            searchParams.delete(URLParams.PriceMax);
            searchParams.append(URLParams.PriceMax, inputValue.min);
          }
          if(productsByPrice[0].price > Number(inputValue.max)) {
            setInputValue((prev) => ({...prev, max: String(productsByPrice[0].price)}));
            searchParams.delete(URLParams.PriceMax);
            searchParams.append(URLParams.PriceMax, String(productsByPrice[0].price));
          }
          if(!productsByPrice.find((product) => product.price === Number(inputValue.max)) && productsByPrice[0].price < Number(inputValue.max)) {
            const newPriceInput = productsByPrice.find((product, i) => {
              const nextProduct = productsByPrice[i + 1];
              return nextProduct && (product.price < Number(inputValue.max)) && (nextProduct.price > Number(inputValue.max));
            });
            const newMaxPriceInput = newPriceInput?.price;
            setInputValue((prev) => ({...prev, max: String(newMaxPriceInput)}));
            searchParams.delete(URLParams.PriceMax);
            searchParams.append(URLParams.PriceMax, String(newMaxPriceInput));
          }
        }
      }

      else {
        const piceMaxParam = searchParams.get(URLParams.PriceMax) || '';
        if (piceMaxParam !== inputValue.max) {
          setInputValue((prev) => ({...prev, max: piceMaxParam}));
        }
      }
      setSearchParams(searchParams);
    },
    [debouncedMaxInput, searchParams]
  );

  return(
    <fieldset className="catalog-filter__block">
      <legend className="title title--h5">Цена, ₽</legend>
      <div className="catalog-filter__price-range">
        <div className="custom-input">
          <label>
            <input type="number" id='minPrice' value={inputValue.min} name="price" placeholder={productsByPrice[0] ? productsByPrice[0].price?.toString() : 'от'} onChange={(evt) => setInputValue((prev) => ({...prev, min: evt.target.value}))}/>
          </label>
        </div>
        <div className="custom-input">
          <label>
            <input type="number" id='maxPrice' value={inputValue.max} name="priceUp" placeholder={productsByPrice[0] ? productsByPrice[productsByPrice.length - 1].price?.toString() : 'до'} onChange={(evt) => setInputValue((prev) => ({...prev, max: evt.target.value}))}/>
          </label>
        </div>
      </div>
    </fieldset>
  );
}

