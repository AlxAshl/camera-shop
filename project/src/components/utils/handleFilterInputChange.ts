import { ActionCreatorWithNonInferrablePayload } from '@reduxjs/toolkit';
import { store } from '../../store/store';

export const handleFilterInputChange = (filterName: string, filter: string[], action: ActionCreatorWithNonInferrablePayload) => {
  const mutableFilter = Object.assign([], filter);
  if(filter.includes(filterName)) {
    const updatedFilter = mutableFilter.filter((entry) => entry !== filterName);
    store.dispatch(action(updatedFilter));
  }
  else {
    mutableFilter.push(filterName);
    store.dispatch(action(mutableFilter));
  }
};
