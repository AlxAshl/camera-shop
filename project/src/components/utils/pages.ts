import { PAGE_LIMIT } from '../../const';

function getPagesCount(cameraCount: number): number {
  return Math.ceil(cameraCount / PAGE_LIMIT);
}

export default getPagesCount;
