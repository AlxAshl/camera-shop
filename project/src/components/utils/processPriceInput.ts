import { ActionCreatorWithNonInferrablePayload } from '@reduxjs/toolkit';
import { store } from '../../store/store';
import { ProductType } from '../../types/product';

export const setMinInput = (input: string, productsArray: ProductType[], action: ActionCreatorWithNonInferrablePayload) => {
  if(productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    store.dispatch(action([String(input)]));
  }
  if(!productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    let newMinPriceInput;
    const newMin = productsArray.find((product) => product.price <= Number(input));
    newMin?.price === undefined
      ? newMinPriceInput = productsArray[productsArray.length - 1].price
      : newMinPriceInput = newMin?.price;
    store.dispatch(action([String(newMinPriceInput)]));
  }
};

export const setMaxInput = (input: string, productsArray: ProductType[], action: ActionCreatorWithNonInferrablePayload) => {
  if(productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    store.dispatch(action([String(input)]));
  }
  if(!productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    let newMaxPriceInput;
    const newMax = productsArray.find((product) => product.price >= Number(input));
    newMax?.price === undefined
      ? newMaxPriceInput = productsArray[productsArray.length - 1].price
      : newMaxPriceInput = newMax?.price;
    store.dispatch(action([String(newMaxPriceInput)]));
  }
};
