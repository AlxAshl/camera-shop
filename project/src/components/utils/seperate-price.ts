export function seperatePrice(priceValue: number) {
  let returnedPrice = String(priceValue);
  if(returnedPrice.length > 6) {
    returnedPrice = `${returnedPrice.slice(0, 1)} ${returnedPrice.slice(1, 4)} ${returnedPrice.slice(-3)}`;
  }
  if (returnedPrice.length > 3) {
    returnedPrice = `${returnedPrice.slice(-0, -3)} ${returnedPrice.slice(-3)}`;
  }
  return returnedPrice;
}
