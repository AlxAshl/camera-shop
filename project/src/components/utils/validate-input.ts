export function validateInput(target: (EventTarget & HTMLInputElement) | (EventTarget & HTMLTextAreaElement)) {

  if (target.parentElement?.parentElement && target.name === 'rating' && (Number(target.value) > 0)) {
    if (!target.parentElement.parentElement.parentElement?.className.includes('is-valid')) {
      target.parentElement.parentElement.parentElement?.classList.add('is-valid');
      target.parentElement.parentElement.parentElement?.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && target.name !== 'review' && typeof(target.value) === 'string') {
    if (!target.parentElement.parentElement.className.includes('is-valid') && target.value.length > 0) {
      target.parentElement.parentElement.className += ' is-valid';
      target.parentElement.parentElement.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && typeof(target.value) === 'string'){
    if(target.value.length === 0) {
      target.parentElement.parentElement.className += ' is-invalid';
      target.parentElement.parentElement.classList.remove('is-valid');
    }
  }

  if (target.parentElement?.parentElement && typeof(target.value) === 'string') {
    if (!target.parentElement.parentElement.className.includes('is-valid') && target.value.length >= 5) {
      target.parentElement.parentElement.className += ' is-valid';
      target.parentElement.parentElement.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && target.name === 'review' && target.value.length < 5) {
    if (!target.parentElement.parentElement.className.includes('is-invalid')) {
      target.parentElement.parentElement.className += ' is-invalid';
      target.parentElement.parentElement.classList.remove('is-valid');
    }
  }

}
