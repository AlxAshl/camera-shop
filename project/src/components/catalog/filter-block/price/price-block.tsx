import { useEffect, useState } from 'react';
import { useAppSelector } from '../../../../hooks/useAppSelector';
import { getFilters, getProductsByPrice, getProductsByPriceDesc } from '../../../../store/filters-process/selectors';
import useInputEventListener from '../../../../hooks/use-input-event-listener';
import { maxPriceFilterSetter, minPriceFilterSetter } from '../../../../store/filters-process/filters-process';
import { useSelector } from 'react-redux';
import { setMaxInput, setMinInput } from '../../../utils/processPriceInput';
import { useAppDispatch } from '../../../../hooks/useAppDispatch';


export function PriceBlock() {

  const [inputValue, setInputValue] = useState({ min: '', max: ''});
  const [isInitial, setIsInitial] = useState(true);
  const productsByPriceAsc = useAppSelector(getProductsByPrice);
  const productsByPriceDesc = useAppSelector(getProductsByPriceDesc);
  const {PriceMax, PriceMin} = useSelector(getFilters);
  const dispatch = useAppDispatch();
  const minPriceFiled = document.querySelector('#minPrice');
  const maxPriceFiled = document.querySelector('#maxPrice');
  const returnMinInput = useInputEventListener(minPriceFiled as HTMLInputElement, minPriceFilterSetter);
  const returnMaxInput = useInputEventListener(maxPriceFiled as HTMLInputElement, maxPriceFilterSetter);

  useEffect(
    () => {
      (Number(returnMinInput) > Number(inputValue.max) && inputValue.max !== '') && !isInitial
        ? dispatch(minPriceFilterSetter([inputValue.max]))
        : setMinInput(returnMinInput, productsByPriceDesc, minPriceFilterSetter);
    },
    [returnMinInput]
  );
  useEffect(
    () => {
      Number(returnMaxInput) < Number(inputValue.min) && !isInitial
        ? dispatch(maxPriceFilterSetter([inputValue.min]))
        : setMaxInput(returnMaxInput, productsByPriceAsc, maxPriceFilterSetter);
    },
    [returnMaxInput]
  );

  useEffect(()=> {
    if(PriceMax) {
      setInputValue((prev) => ({...prev, max: String(PriceMax)}));
    }
    if(PriceMin) {
      setInputValue((prev) => ({...prev, min: String(PriceMin)}));
    }
    setIsInitial(false);
  },[isInitial, PriceMax, PriceMin]);

  return(
    <fieldset className="catalog-filter__block">
      <legend className="title title--h5">Цена, ₽</legend>
      <div className="catalog-filter__price-range">
        <div className="custom-input">
          <label>
            <input type="number" data-testid='minPrice-test' id='minPrice' value={inputValue.min} name="price" placeholder={productsByPriceAsc[0] ? productsByPriceAsc[0].price?.toString() : 'от'} onChange={(evt) => setInputValue((prev) => ({...prev, min: evt.target.value}))}/>
          </label>
        </div>
        <div className="custom-input">
          <label>
            <input type="number" data-testid='maxPrice-test' id='maxPrice' value={inputValue.max} name="priceUp" placeholder={productsByPriceAsc[0] ? productsByPriceAsc[productsByPriceAsc.length - 1].price?.toString() : 'до'} onChange={(evt) => setInputValue((prev) => ({...prev, max: evt.target.value}))}/>
          </label>
        </div>
      </div>
    </fieldset>
  );
}

