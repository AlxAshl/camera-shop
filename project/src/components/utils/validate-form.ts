import { reviewDataType } from '../../types/review';
import { validateInput } from './validate-input';


export function validateForm(data: reviewDataType) {
  let validationStatus = true;
  Object.entries(data).forEach(([key, value]) =>{
    if (key !== 'review') {
      if(value === '') {
        const failedInput = document.querySelector(`input[name=${key}]`);
        validateInput(failedInput as HTMLInputElement);
        validationStatus = false;
      }
    }
    if(key === 'review') {
      if(String(value).length < 5) {
        const failedInput = document.querySelector(`textarea[name=${key}]`);
        validateInput(failedInput as HTMLInputElement);
      }
    }
    if(key === 'rating') {
      if(value < 1) {
        const failedInput = document.querySelector(`input[name=${key}]`);
        if(failedInput && failedInput.parentElement?.parentElement){
          failedInput.parentElement.parentElement.parentElement?.classList.add('is-invalid');
        }
        validationStatus = false;
      }
    }
  });
  return validationStatus;
}

