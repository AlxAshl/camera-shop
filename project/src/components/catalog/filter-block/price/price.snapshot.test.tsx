import { renderWithProviders } from '../../../../test/utils/render-with-redux';
import { Price } from './price';

test('Render Price', () => {
  const { asFragment } = renderWithProviders(<Price />);
  expect(asFragment()).toMatchSnapshot();
});
