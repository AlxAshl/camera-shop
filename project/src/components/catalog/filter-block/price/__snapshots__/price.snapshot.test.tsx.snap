// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Price 1`] = `
<DocumentFragment>
  <fieldset
    class="catalog-filter__block"
  >
    <legend
      class="title title--h5"
    >
      Цена, ₽
    </legend>
    <div
      class="catalog-filter__price-range"
    >
      <div
        class="custom-input"
      >
        <label>
          <input
            data-testid="minPrice-test"
            id="minPrice"
            name="price"
            placeholder="от"
            type="number"
            value=""
          />
        </label>
      </div>
      <div
        class="custom-input"
      >
        <label>
          <input
            data-testid="maxPrice-test"
            id="maxPrice"
            name="priceUp"
            placeholder="до"
            type="number"
            value=""
          />
        </label>
      </div>
    </div>
  </fieldset>
</DocumentFragment>
`;
